controllers/contact-controller.js

```diff
@@ -308,11 +321,24 @@ const ContactController = Contact => {
     };
 
     const contact = await Contact.create(contactRecord);
+
+    // Link all of the csam reports related to this contact
+    if (body.csamReports && body.csamReports.length) {
+      await Promise.all(
+        body.csamReports.map(r =>
+          CSAMReportController.connectToContact(contact.id, r.id, accountSid),
+        ),
+      );
+    }
+
     return contact;
```

Comment:
Couple things to note:
1. Notice that when we return `contact`, we won't be including the CSAM reports, since we added them afterward.
2. Perhaps it's time we gave transactions a try. This has been in our backlog for a long time [CHI-434](https://bugs.benetech.org/browse/CHI-434). See the Managed Transaction section of the [docs](https://sequelize.org/master/manual/transactions.html). The value of a transaction is that you get consistency -- everything updates or nothing updates. We don't have to worry about an error scenario where the update to `Contact` succeeds but the update to `CSAMReports` fails.

You could try something like:
```
  const contact = await sequelize.transaction(async (t) => {

    const newContact = await Contact.create(contactRecord, { transaction: t });

    await CSAMReports.update(
      { contactId: newContact.id },
      { 
        where:  { id: csamReports, accountId: accountSid },
        transaction: t 
      });

    return newContact;

  });

  return contact;
```

To be clear -- I haven't tested that and I'm not sure it works.

I think it's worth giving transactions a shot here. If you keep running into trouble with it, we can abandon and save  CHI-434 for later.

I also notice that the transaction queries still doesn't solve the issue that we're returning the contact without the CSAM reports attached. You may have better ideas on how to do that than me.

Reply:
By changing the approach to a bulk update I think the create contact method won't fail anymore, it just won't link if there are "broken report ids". Do you think is still worth doing this if this does not solves the association?

Reply:
The data could still potentially be left in an inconsistent state, even with the single update query. If the contact is created, but the update query to associate the CSAM reports fails, you will be left with a Contact with no CSAM reports. If there is a transaction wrapping both queries, it will be 'all or nothing', either a Contact with CSAM associations, or no Contact at all - which may be better because it can just be retried in full and doesn't leave potentially confusing, incomplete data on the system in the meantime.